File: crates/trident-acl-agent-tester/src/scenario.rs
//! Scenario model and report types for the RP proxy runner.

use std::{collections::BTreeMap, path::Path, time::Duration};

use anyhow::Context;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Scenario {
    pub steps: Vec<ScenarioStep>,
}

impl Scenario {
    pub fn load(path: &Path) -> anyhow::Result<Self> {
        let contents = std::fs::read_to_string(path)
            .with_context(|| format!("failed to read scenario {}", path.display()))?;
        serde_yaml::from_str(&contents).context("failed to parse scenario yaml")
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum ScenarioStep {
    Patch {
        patch: PatchStep,
    },
    Expect {
        expect: ExpectStep,
    },
    AssertFailureReason {
        #[serde(rename = "assert-failure-reason")]
        assert_failure_reason: String,
    },
}

#[derive(Debug, Clone, Deserialize)]
pub struct PatchStep {
    pub request: Option<String>,
    #[serde(rename = "request-id")]
    pub request_id: Option<String>,
    #[serde(rename = "target-os-image-version")]
    pub target_os_image_version: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ExpectStep {
    pub state: String,
    #[serde(rename = "observed-request-id")]
    pub observed_request_id: Option<String>,
    #[serde(default = "default_timeout")]
    pub timeout: DurationDef,
    #[serde(default)]
    pub expect_timeout: bool,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(transparent)]
pub struct DurationDef(#[serde(deserialize_with = "deserialize_duration")] pub Duration);

impl Default for DurationDef {
    fn default() -> Self {
        Self(default_timeout().0)
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ScenarioReport {
    pub passed: bool,
    pub steps: Vec<StepReport>,
}

#[derive(Debug, Clone, Serialize)]
pub struct StepReport {
    pub index: usize,
    pub kind: String,
    pub passed: bool,
    pub elapsed_ms: u128,
    pub message: String,
    pub expected: Option<serde_json::Value>,
    pub actual: Option<serde_json::Value>,
}

fn default_timeout() -> DurationDef {
    DurationDef(Duration::from_secs(60))
}

fn deserialize_duration<'de, D>(deserializer: D) -> Result<Duration, D::Error>
where
    D: serde::Deserializer<'de>,
{
    let raw = String::deserialize(deserializer)?;
    humantime::parse_duration(&raw).map_err(serde::de::Error::custom)
}

pub fn labels_from_patch_step(step: &PatchStep) -> BTreeMap<String, String> {
    let mut labels = BTreeMap::new();
    if let Some(request) = &step.request {
        labels.insert(
            "kubernetes.azure.com/trident-abupdate-request".to_string(),
            request.clone(),
        );
    }
    if let Some(request_id) = &step.request_id {
        labels.insert(
            "kubernetes.azure.com/trident-abupdate-request-id".to_string(),
            request_id.clone(),
        );
    }
    if let Some(target) = &step.target_os_image_version {
        labels.insert(
            "kubernetes.azure.com/trident-abupdate-target-os-image-version".to_string(),
            target.clone(),
        );
    }
    labels
}

File: tools/storm/aclagent/proxies/scenario.go
package proxies

import (
	"fmt"
	"os"
	"time"

	"gopkg.in/yaml.v3"
)

type Scenario struct {
	Steps []ScenarioStep `yaml:"steps"`
}

type ScenarioStep struct {
	Patch  *PatchStep  `yaml:"patch,omitempty"`
	Expect *ExpectStep `yaml:"expect,omitempty"`
}

type PatchStep struct {
	NodeUpdateID         string `yaml:"node-update-id,omitempty" json:"nodeUpdateId,omitempty"`
	OperationID          string `yaml:"operation-id,omitempty" json:"operationId,omitempty"`
	Operation            string `yaml:"operation,omitempty" json:"operation,omitempty"`
	TargetOSImageVersion string `yaml:"target-os-image-version,omitempty" json:"targetVersion,omitempty"`
}

type ExpectStep struct {
	OperationID   string        `yaml:"operation-id,omitempty" json:"operationId,omitempty"`
	Operation     string        `yaml:"operation,omitempty" json:"operation,omitempty"`
	Code          string        `yaml:"code" json:"code"`
	Timeout       time.Duration `yaml:"-" json:"timeoutSeconds"`
	TimeoutRaw    string        `yaml:"timeout,omitempty" json:"-"`
	ExpectTimeout bool          `yaml:"expect-timeout,omitempty" json:"expectTimeout,omitempty"`
}

type ScenarioReport struct {
	Passed bool         `json:"passed"`
	Steps  []StepReport `json:"steps"`
}

type StepReport struct {
	Index     int    `json:"index"`
	Kind      string `json:"kind"`
	Passed    bool   `json:"passed"`
	ElapsedMS int64  `json:"elapsedMs"`
	Message   string `json:"message"`
	Expected  any    `json:"expected,omitempty"`
	Actual    any    `json:"actual,omitempty"`
}

func LoadScenario(path string) (*Scenario, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read scenario %s: %w", path, err)
	}
	var scenario Scenario
	if err := yaml.Unmarshal(data, &scenario); err != nil {
		return nil, fmt.Errorf("failed to parse scenario yaml: %w", err)
	}
	if err := scenario.Validate(); err != nil {
		return nil, err
	}
	return &scenario, nil
}

func (s *Scenario) Validate() error {
	for index := range s.Steps {
		step := &s.Steps[index]
		kinds := 0
		if step.Patch != nil {
			kinds++
		}
		if step.Expect != nil {
			kinds++
		}
		if kinds != 1 {
			return fmt.Errorf("scenario step %d must set exactly one of patch/expect", index)
		}
		if step.Expect != nil {
			timeout := 60 * time.Second
			if step.Expect.TimeoutRaw != "" {
				var err error
				timeout, err = time.ParseDuration(step.Expect.TimeoutRaw)
				if err != nil {
					return fmt.Errorf("scenario step %d has invalid timeout %q: %w", index, step.Expect.TimeoutRaw, err)
				}
			}
			step.Expect.Timeout = timeout
		}
	}
	return nil
}
